pkg/ext-proc/test/hermetic_test.go

```diff
@@ -0,0 +1,155 @@
+		pods        []*backend.PodMetrics
+		wantHeaders []*configPb.HeaderValueOption
+		wantBody    []byte
+		wantErr     bool
+	}{
+		{
+			name: "success",
+			req:  GenerateRequest("my-model"),
+			// pod-1 will be picked because it has relatively low queue size, with the requested
+			// model being active, and has low KV cache.
+			pods: []*backend.PodMetrics{
+				{
+					Pod: FakePod(0),
+					Metrics: backend.Metrics{
+						WaitingQueueSize:    0,
+						KVCacheUsagePercent: 0.2,
+						ActiveModels: map[string]int{
+							"foo": 1,
+							"bar": 1,
+						},
+					},
+				},
+				{
+					Pod: FakePod(1),
+					Metrics: backend.Metrics{
+						WaitingQueueSize:    3,
+						KVCacheUsagePercent: 0.1,
+						ActiveModels: map[string]int{
+							"foo":      1,
+							"my-model": 1,
+						},
+					},
+				},
+				{
+					Pod: FakePod(2),
+					Metrics: backend.Metrics{
+						WaitingQueueSize:    10,
+						KVCacheUsagePercent: 0.2,
+						ActiveModels: map[string]int{
+							"foo": 1,
+						},
+					},
+				},
+			},
+			wantHeaders: []*configPb.HeaderValueOption{
+				{
+					Header: &configPb.HeaderValue{
+						Key:      "target-pod",
+						RawValue: []byte("address-1"),
+					},
+				},
+			},
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			client, cleanup := setUpServer(t, test.pods)
+			t.Cleanup(cleanup)
+			want := &extProcPb.ProcessingResponse{
+				Response: &extProcPb.ProcessingResponse_RequestBody{
+					RequestBody: &extProcPb.BodyResponse{
+						Response: &extProcPb.CommonResponse{
+							HeaderMutation: &extProcPb.HeaderMutation{
+								SetHeaders: test.wantHeaders,
+							},
+							// TODO: Also check body once it's added.
+							// BodyMutation: &extProcPb.BodyMutation{
+							// 	Mutation: &extProcPb.BodyMutation_Body{
+							// 		Body: test.wantBody,
+							// 	},
+							// },
+						},
+					},
+				},
+			}
+			res, err := sendRequest(t, client, test.req)
+
+			if (err != nil) != test.wantErr {
+				t.Fatalf("Unexpected error, got %v, want %v", err, test.wantErr)
+			}
+
+			if diff := cmp.Diff(want, res, protocmp.Transform()); diff != "" {
+				t.Errorf("Unexpected response, (-want +got): %v", diff)
+			}
+		})
+	}
+
+}
+
+func setUpServer(t *testing.T, pods []*backend.PodMetrics) (client extProcPb.ExternalProcessor_ProcessClient, cleanup func()) {
+	server := StartExtProc(port, time.Second, time.Second, pods)
+
+	address := fmt.Sprintf("localhost:%v", port)
+	// Create a grpc connection
+	conn, err := grpc.NewClient(address, grpc.WithTransportCredentials(insecure.NewCredentials()))
+	if err != nil {
+		log.Fatalf("Failed to connect to %v: %v", address, err)
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	client, err = extProcPb.NewExternalProcessorClient(conn).Process(ctx)
+	if err != nil {
+		log.Fatalf("Failed to create client: %v", err)
+	}
+	return client, func() {
+		cancel()
+		conn.Close()
+		server.GracefulStop()
```

Comment:
is this going to kill the go routine that started the server here https://github.com/kubernetes-sigs/llm-instance-gateway/pull/29/files#diff-8dd426243c10996d6a7c647c65b54d42d7a962c9be63e45fe8253e9a2612a2c7R48 ?

Reply:
yes